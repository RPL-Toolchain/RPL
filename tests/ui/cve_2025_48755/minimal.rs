//@revisions: normal
//@[normal]compile-flags: -Z inline-mir=false
use std::alloc::{Layout, alloc, alloc_zeroed, dealloc};

pub fn alloc_maybe_zero(sz_u64: usize) {
    let sz_bytes = sz_u64 * size_of::<u64>();
    let layout = Layout::from_size_align(sz_bytes, 8).unwrap();
    let ptr = unsafe { alloc(layout) };
    //~^ERROR: public function `alloc_maybe_zero` allocates a pointer that may be zero-sized, which is an undefined behavior
    //~|NOTE:  See https://doc.rust-lang.org/std/alloc/fn.alloc.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.alloc
    //~|NOTE:  `#[deny(rpl::alloc_maybe_zero)]` on by default
    unsafe { dealloc(ptr, layout) }
}

pub fn alloc_zeroed_maybe_zero(sz_u64: usize) {
    let sz_bytes = sz_u64 * size_of::<u64>();
    let layout = Layout::from_size_align(sz_bytes, 8).unwrap();
    let ptr = unsafe { alloc_zeroed(layout) };
    //~^ERROR: public function `alloc_zeroed_maybe_zero` allocates a pointer that may be zero-sized, which is an undefined behavior
    //~|NOTE:  See https://doc.rust-lang.org/std/alloc/fn.alloc_zeroed.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.alloc_zeroed
    unsafe { dealloc(ptr, layout) }
}
