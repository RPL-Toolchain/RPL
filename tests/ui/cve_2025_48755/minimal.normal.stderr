error: public function `alloc_maybe_zero` allocates a pointer that may be zero-sized, which is an undefined behavior
  --> tests/ui/cve_2025_48755/minimal.rs:8:24
   |
LL | pub fn alloc_maybe_zero(sz_u64: usize) {
   |                         ------ size passed in here
...
LL |     let ptr = unsafe { alloc(layout) };
   |                        ^^^^^^^^^^^^^ pointer created here
   |
   = note: See https://doc.rust-lang.org/std/alloc/fn.alloc.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.alloc
   = note: `#[deny(rpl::alloc_maybe_zero)]` on by default

error: public function `alloc_zeroed_maybe_zero` allocates a pointer that may be zero-sized, which is an undefined behavior
  --> tests/ui/cve_2025_48755/minimal.rs:18:24
   |
LL | pub fn alloc_zeroed_maybe_zero(sz_u64: usize) {
   |                                ------ size passed in here
...
LL |     let ptr = unsafe { alloc_zeroed(layout) };
   |                        ^^^^^^^^^^^^^^^^^^^^ pointer created here
   |
   = note: See https://doc.rust-lang.org/std/alloc/fn.alloc_zeroed.html and https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#method.alloc_zeroed

error: aborting due to 2 previous errors

